### PR TITLE
update several minor in conjugate graph

### PR DESCRIPTION
**Pull request by ShawnShawnYou**
This PR contains several minor changes in conjugate graph: 
* support int8 in conjugate graph
* support using conjugate graph in fresh_hnsw
* add multi-thread test
* modify search logic with conjugate graph
* add maximum degree limit in conjugate graph

### CODE REVIEW

File: src/index/hnsw.cpp [hunk 1]
```diff
@@ -273,9 +277,13 @@ HNSW::knn_search(const DatasetPtr& query,
                 return this->alg_hnsw_->getDistanceByLabel(label, vector);
             };
             conjugate_graph_->EnhanceResult(results, func);
+            k = original_k;
         }
 
         // return result
+        while (results.size() > k) {
+            results.pop();
+        }
```
Comment (ShawnShawnYou) [inline src/index/hnsw.cpp:277]: comparing with original search process, here only move partial pop operations from hnswalg to hnsw. So it wouldn't affect the search speed.

File: src/index/hnsw.cpp [hunk 2]
```diff
@@ -238,10 +238,14 @@ HNSW::knn_search(const DatasetPtr& query,
...
 
         // perform search
+        int64_t original_k = k;
         std::priority_queue<std::pair<float, size_t>> results;
         double time_cost;
         try {
             Timer t(time_cost);
+            if (use_conjugate_graph_ and params.use_conjugate_graph_search) {
+                k = LOOK_AT_K;
+            }
```
Comment (ShawnShawnYou) [inline src/index/hnsw.cpp:239]: here adjust k to provide enough information for conjugate graph to enhance result

File: src/impl/conjugate_graph.cpp [hunk 3]
```diff
@@ -54,6 +57,10 @@ ConjugateGraph::get_neighbors(int64_t from_tag_id) const {
 tl::expected<uint32_t, Error>
 ConjugateGraph::EnhanceResult(std::priority_queue<std::pair<float, size_t>>& results,
                               const std::function<float(int64_t)>& distance_of_tag) const {
+    if (this->is_empty()) {
+        return 0;
+    }
+
```
Comment (ShawnShawnYou) [inline src/impl/conjugate_graph.cpp:63]: here allow use conjugate graph as default setting without influencing search process

File: src/impl/conjugate_graph.cpp [hunk 4]
```diff
@@ -27,6 +27,9 @@ ConjugateGraph::AddNeighbor(int64_t from_tag_id, int64_t to_tag_id) {
         return false;
     }
     auto& neighbor_set = conjugate_graph_[from_tag_id];
+    if (neighbor_set.size() >= MAXIMUM_DEGREE) {
+        return false;
+    }
```
Comment (ShawnShawnYou) [inline src/impl/conjugate_graph.cpp:32]: here limit out degree to limit graph size and enhance latency

File: src/index/hnsw.cpp [hunk 5]
```diff
@@ -238,10 +238,14 @@ HNSW::knn_search(const DatasetPtr& query,
         auto params = HnswSearchParameters::FromJson(parameters);
 
         // perform search
+        int64_t original_k = k;
         std::priority_queue<std::pair<float, size_t>> results;
         double time_cost;
         try {
             Timer t(time_cost);
+            if (use_conjugate_graph_ and params.use_conjugate_graph_search) {
+                k = LOOK_AT_K;
```
Comment (inabao) [inline src/index/hnsw.cpp]: consider the situation when `k > LOOK_AT_K`

File: src/index/hnsw.cpp [hunk 6]
```diff
@@ -701,10 +709,15 @@ HNSW::brute_force(const DatasetPtr& query, int64_t k) {
         float* dists = (float*)allocator_->Allocate(sizeof(float) * k);
         result->Distances(dists);
 
-        auto vector = query->GetFloat32Vectors();
+        const void* vector;
+        if (type_ == DataTypes::DATA_TYPE_INT8) {
```
Comment (inabao) [inline src/index/hnsw.cpp]: use `get_vectors` here to replace the judgement.

File: src/index/hnsw.cpp [hunk 7]
```diff
@@ -739,18 +752,37 @@ HNSW::pretrain(const std::vector<int64_t>& base_tag_ids,
...
 
+    uint32_t data_size = 0;
     uint32_t add_edges = 0;
     int64_t topk_neighbor_tag_id;
-    const float* topk_data;
-    std::shared_ptr<float[]> generated_data(new float[dim_]);
+    const int8_t* topk_data;
     auto base = Dataset::Make();
     auto generated_query = Dataset::Make();
+    if (type_ == DataTypes::DATA_TYPE_INT8) {
```
Comment (inabao) [inline src/index/hnsw.cpp:779]: ditto

File: src/index/hnsw.cpp [hunk 8]
```diff
@@ -739,18 +752,37 @@ HNSW::pretrain(const std::vector<int64_t>& base_tag_ids,
...
+            ->Owner(false);
+    }
+
     base->Dim(dim_)->NumElements(1)->Owner(false);
-    generated_query->Dim(dim_)->NumElements(1)->Float32Vectors(generated_data.get())->Owner(false);
 
     for (const int64_t& base_tag_id : base_tag_ids) {
         try {
-            base->Float32Vectors(this->alg_hnsw_->getDataByLabel(base_tag_id));
+            if (type_ == DataTypes::DATA_TYPE_INT8) {
```
Comment (inabao) [inline src/index/hnsw.cpp]: ditto

File: src/index/hnsw.cpp [hunk 9]
```diff
@@ -776,11 +808,18 @@ HNSW::pretrain(const std::vector<int64_t>& base_tag_ids,
             if (topk_neighbor_tag_id == base_tag_id) {
                 continue;
             }
-            topk_data = this->alg_hnsw_->getDataByLabel(topk_neighbor_tag_id);
+            topk_data = (const int8_t*)this->alg_hnsw_->getDataByLabel(topk_neighbor_tag_id);
 
             for (int d = 0; d < dim_; d++) {
-                generated_data.get()[d] = vsag::GENERATE_OMEGA * base->GetFloat32Vectors()[d] +
-                                          (1 - vsag::GENERATE_OMEGA) * topk_data[d];
+                if (type_ == DataTypes::DATA_TYPE_INT8) {
```
Comment (inabao) [inline src/index/hnsw.cpp:820]: ditto

File: src/index/hnsw.cpp (hunk 6 above)
Comment (ShawnShawnYou) [reply src/index/hnsw.cpp]: done

File: src/index/hnsw.cpp [hunk 10]
```diff
@@ -824,4 +850,18 @@ HNSW::get_vectors(const vsag::DatasetPtr& base, void** vectors_ptr, size_t* data
...
 
+void
+HNSW::set_dataset(const DatasetPtr& base, const void* vectors_ptr, uint32_t num_element) const {
+    if (type_ == DataTypes::DATA_TYPE_FLOAT) {
+        base->Float32Vectors((float*)vectors_ptr)
+            ->Dim(dim_)
+            ->Owner(false)
+            ->NumElements(num_element);
+    } else if (type_ == DataTypes::DATA_TYPE_INT8) {
+        base->Int8Vectors((int8_t*)vectors_ptr)->Dim(dim_)->Owner(false)->NumElements(num_element);
```
Comment (jiaweizone) [inline src/index/hnsw.cpp:880]: indent

File: src/index/hnsw.cpp [hunk 11]
```diff
@@ -824,4 +850,18 @@ HNSW::get_vectors(const vsag::DatasetPtr& base, void** vectors_ptr, size_t* data
...
+    } else {
+        throw std::invalid_argument(fmt::format("no support for this metric: {}", (int)type_));
```
Comment (jiaweizone) [inline src/index/hnsw.cpp]: no support for this data type: ?

**Review by jiaweizone — APPROVED** [review]
LGTM

**Review by wxyucs — APPROVED** [review]
lgtm

**Pull request merged**